connect/runtime/src/main/java/org/apache/kafka/connect/storage/OffsetUtils.java

```diff
@@ -53,4 +63,64 @@ public static <K, V> void validateFormat(Map<K, V> offsetData) {
                 throw new DataException("Offsets may only contain primitive types as values, but field " + entry.getKey() + " contains " + schemaType);
         }
     }
+
+    /**
+     * Parses a partition key that is read back from an offset backing store and add / remove the partition in the
+     * provided {@code connectorPartitions} map. If the partition key has an unexpected format, a warning log is emitted
+     * and nothing is added / removed in the {@code connectorPartitions} map.
+     * @param partitionKey the partition key to be processed
+     * @param offsetValue the offset value corresponding to the partition key; determines whether the partition should
+     *                    be added to the {@code connectorPartitions} map or removed depending on whether the offset
+     *                    value is null or not.
+     * @param keyConverter the key converter to deserialize the partition key
+     * @param connectorPartitions the map from connector names to its set of partitions which needs to be updated after
+     *                            processing
+     */
+    @SuppressWarnings("unchecked")
+    public static void processPartitionKey(byte[] partitionKey, byte[] offsetValue, Converter keyConverter,
+                                           Map<String, Set<Map<String, Object>>> connectorPartitions) {
+
+        // The key is expected to always be of the form [connectorName, partition] where connectorName is a
+        // string value and partition is a Map<String, Object>
+
+        if (partitionKey == null) {
+            log.warn("Ignoring offset partition key with an unexpected null value");
+            return;
+        }
+        // The topic parameter is irrelevant for the JsonConverter which is the internal converter used by
+        // Connect workers.
+        Object deserializedValue = keyConverter.toConnectData("", partitionKey).value();
+        if (!(deserializedValue instanceof List)) {
```

Comment:
Worth adding logic to explicitly handle null? I know that it'll be implicitly handled by this condition but the message `Expected type: java.util.List, actual type: null` kind of looks like an error in the logging logic and may confuse users if they see that.

Reply:
I think explicitly declaring that the value is null is the best thing here. It's probably obvious to both of us what that log line would mean now, but we've already looked at this code. If a user sees this they may not know, or at least doubt, what this means and waste time locating this log message in the code base to check what it means.

The issue with `NullType` is that this message would still be misleading; sure, in a theoretical sense, `null` is a type, but it's most-commonly used as a value, and people seeing errors about actual vs. expected type could be confused if they're only familiar with the latter usage.

Reply:
Wait, never mind--we already handle that by null-checking the `partitionKey` argument. The converter turns nulls into other nulls so anything here for `deserializedKey` would be redundant.